### PR TITLE
KE2: Upgrade to Kotlin 2.1.0; restore basic type parameter and type argument extraction

### PR DESCRIPTION
**Pull request by smowton**
**Kotlin 2.1.0**

Needed to upgrade this due to finding compiler crashes when dealing with Java classes that declare type parameters against constructors.

Two snags with this:

1. Kotlin analysis API now requires Caffeine, a caching framework, but this isn't a dependency of either of the kotlin packages we depend on (guess: they're intended to run in a context that already has caffeine on the classpath -- perhaps IntelliJ already does, for example?), and I can't find an appropriate Kotlin package to retarget. Just adding it as an extra dependency suffices for now, but if there is a more appropriate Maven dependency to target that will pull it in for us, we should do that.
2. Probably relatedly, since Caffeine schedules background work to clean up caches periodically, the extractor no longer terminates without a hard `System.exit(0)` once all our work is done. Ideally we'd work out (a) is it actually Caffeine keeping the JVM awake (snork), and (b) if it is, how should it be asked to make an orderly shutdown.

**Generics extraction**

Done here:
* Enabled extraction of type parameters
* Enabled using type arguments in TRAP labels and function signatures
* Enabled extracting wildcard types

Not done yet:
* Actually extract specialised classes (e.g. `Map<String, String>`) -- requires re-enabling the logic to emit a class outline into the current trap file, and type substitution
* File classes / other special class names, like companion objects and such
* Any of the Java/Kotlin equivalence stuff -- so `f(Int, Float?)` will extract is `kotlin.Int, kotlin.Float` not `int, java.lang.Float` as it ought to for J/K compatibility.

### CODE REVIEW

File: java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt [hunk 1]
```diff
@@ -187,7 +188,7 @@ private fun doAnalysis(
     val ourLanguageVersionSettings = k2args.toLanguageVersionSettings(MessageCollector.NONE)
 
     val session = buildStandaloneAnalysisAPISession {
-        registerProjectService(KotlinLifetimeTokenProvider::class.java, KotlinAlwaysAccessibleLifetimeTokenProvider())
+        registerProjectService(KotlinLifetimeTokenFactory::class.java, KotlinAlwaysAccessibleLifetimeTokenFactory())
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt:191]: Changed in Kotlin 2.1.0

File: java/kotlin-extractor2/src/main/kotlin/KotlinUsesExtractor.kt [hunk 2]
```diff
@@ -252,32 +232,6 @@ open class KotlinUsesExtractor(
                 }
             }
 
-        // The Kotlin compiler internal representation of Outer<A, B>.Inner<C, D>.InnerInner<E,
-        // F>.someFunction<G, H>.LocalClass<I, J> is LocalClass<I, J, G, H, E, F, C, D, A, B>. This
-        // function returns [A, B, C, D, E, F, G, H, I, J].
-        private fun orderTypeArgsLeftToRight(
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/KotlinUsesExtractor.kt:258]: Pleasingly this could die, because the Kotlin Analysis API represents this as `[[A, B], [C, D], [E, F], [G, H], [I, K]]` via the `KaClassType.qualifiers` property, which is a much more useful than the IR representation.

File: java/kotlin-extractor2/src/main/kotlin/entities/Class.kt [hunk 3]
```diff
@@ -214,8 +214,8 @@ fun KotlinUsesExtractor.useClassSource(c: KaClassSymbol): Label<out DbClassorint
  */
 private fun tryReplaceType(
     cBeforeReplacement: KaClassSymbol,
-    argsIncludingOuterClassesBeforeReplacement: List<Nothing>?
-): Pair<KaClassSymbol, List<Nothing>?> {
+    argsIncludingOuterClassesBeforeReplacement: List<List<KaTypeProjection>>?
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/entities/Class.kt:217]: Note this `List<List<KaTypeProjection>>` format occurs quite a lot, and represents the type-args of outer classes, if any.

File: java/kotlin-extractor2/src/main/kotlin/entities/Function.kt [hunk 4]
```diff
@@ -538,7 +544,7 @@ OLD: KE1
         */
         val functionSyntax = f.psi as? KtDeclarationWithBody
         val locId = functionSyntax?.let {
-            tw.getLocation(functionSyntax ?: TODO())
+            tw.getLocation(functionSyntax)
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/entities/Function.kt:547]: Incidental cleanup -- the ?: was unreachable

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeArgument.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            tw.writeHasLocation(wildcardLabel, tw.unknownLocation)
+            tw.getLabelFor<DbTypebound>("@\"bound;0;{$wildcardLabel}\"") {
+                tw.writeTypeBounds(it, boundLabel, 0, wildcardLabel)
+            }
+        }
+
+    // Note this function doesn't return a signature because type arguments are never
+    // incorporated into function signatures.
+    return when (arg) {
+        is KaStarTypeProjection -> {
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/entities/TypeArgument.kt:30]: This is all basically unchanged from the IR format -- the representation of variances is essentially the same. Basic translation to wildcards (`out T` == `? extends T`, `in T` == `? super T`) was likewise restored.

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        when (tp.variance) {
+            Variance.IN_VARIANCE -> addModifiers(id, "in")
+            Variance.OUT_VARIANCE -> addModifiers(id, "out")
+            else -> {}
+        }
+
+        // extractAnnotations(tp, id)
+        // TODO: introduce annotations once they can be disambiguated from bounds, which are
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt:81]: Note this is an old TODO; this was commented out in KE1

File: java/kotlin-extractor2/src/main/kotlin/entities/Types.kt [hunk 7]
```diff
@@ -48,32 +47,22 @@ fun KotlinUsesExtractor.getTypeParameterParentLabel(param: KaTypeParameterType)
...
+                useFunction(it, useDeclarationParentOf(it, true) ?: TODO(), listOf(),  /* OLD: KE1 noReplace = true */)
             else -> {
                 logger.error("Unexpected type parameter parent $it")
                 null
             }
         }
     }
 
-context(KaSession)
-fun KotlinUsesExtractor.getTypeParameterLabel(param: KaTypeParameterType): String {
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/entities/Types.kt:60]: Gone to `TypeParameter.kt`

File: java/kotlin-extractor2/src/main/kotlin/utils/TypeResults.kt [hunk 8]
```diff
@@ -14,21 +13,14 @@ OLD: KE1
...
-*/
 data class TypeResultGeneric<SignatureType, out LabelType : AnyDbType>(
     val id: Label<out LabelType>,
-    /*
-    OLD: KE1
-        val signature: SignatureType?,
-        val shortName: String
-    */
+    // OLD: KE1: val signature: SignatureType?,
+    val shortName: String
```
Comment (smowton) [inline java/kotlin-extractor2/src/main/kotlin/utils/TypeResults.kt:19]: Passing `shortName` around is restored (used to generate names for wildcard types, like "? extends Blah" where "Blah" is the shortName of the underlying type). I haven't looked into how we're using `signature` yet so the relevant arguments are still behind OLD/KE1 comments.

**Comment by ben-manes** [comment]
> Probably relatedly, since Caffeine schedules background work to clean up caches periodically, the extractor no longer terminates without a hard System.exit(0) once all our work is done. Ideally we'd work out (a) is it actually Caffeine keeping the JVM awake (snork), and (b) if it is, how should it be asked to make an orderly shutdown.

Caffeine doesn’t create any threads directly or default to use any non-daemon threads indirectly. It will defers anything that can be async and calls foreign user code to an executor, like eviction and its listener. That by default is ForkJoinPool.commonPool() which uses daemon threads. Users can set a custom pool, many use Runnable::run to execute on the caller since caffeine’s own logic is inexpensive.

You can use jstack to find the blocking thread. If it’s related to the cache then it would be custom configuration.

**Comment by smowton** [comment]
Confirmed the thread in question doesn't belong to Caffeine; rather it is a thread associated with `com.intellij.util.concurrency.AppScheduledExecutorService`, which we indirectly activate via `KaSessionProvider.getInstance` -> `KaFirSessionProvider`'s init block -> `scheduledCacheMaintenance = AppExecutorUtil.getAppScheduledExecutorService().scheduleWithFixedDelay`.

An hour of digging later I am not much closer to identifying what the ownership structure for that is supposed to be, and while there are various shutdown functions I can't see any paths from disposing of standalone analysis sessions or similar that would clean up the thread pool.

Bleeding-edge Kotlin provides https://github.com/JetBrains/kotlin/blob/196ae1b0d009f633000910da5d5e6e50433cdb96/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt#L275 but warns about how calling that shouldn't be necessary. It isn't present in Kotlin 2.1.0, and neither is the destroy function that it calls despite that having been in IntelliJ for years.

In summary, I suggest we shrug and continue to simply hard-kill the JVM when we're done.

File: MODULE.bazel [hunk 9]
```diff
@@ -171,9 +171,9 @@ maven = use_extension("@rules_jvm_external//:extensions.bzl", "maven")
 maven.install(
     name = "maven_deps",
     artifacts = [
-        "org.jetbrains.kotlin:%s:2.0.20-RC2" % kotlin_lib
+        "org.jetbrains.kotlin:%s:2.1.0" % kotlin_lib
         for kotlin_lib in ("kotlin-annotation-processing", "kotlin-compiler")
-    ],
+    ] + [ "com.github.ben-manes.caffeine:caffeine:2.9.3" ]  ,
```
Comment (igfoo) [inline MODULE.bazel:176]: I think it would be worth adding a comment like
```
We need the version specified in https://github.com/JetBrains/kotlin/blob/master/gradle/libs.versions.toml
```
so we don't upgrade to the latest version without thinking.

Depending on what happens with https://youtrack.jetbrains.com/issue/KT-73751/Analysis-API-Caffeine-dependency we might be able to do better here in the future.

File: java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt [hunk 10]
```diff
@@ -166,6 +166,7 @@ OLD: KE1
         )
         dtw.flush()
         loggerBase.close()
+        System.exit(0) // TODO: figure out what's keeping the JVM awake
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt:169]: Filed a ticket for this here: https://youtrack.jetbrains.com/issue/KT-73753/Analysis-API-doesnt-terminate

File: java/kotlin-extractor2/src/main/kotlin/entities/Class.kt [hunk 11]
```diff
@@ -183,24 +183,24 @@ fun KotlinFileExtractor.extractClassSource(
...
+data class ClassLabelResults(val classLabel: String, val shortName: String)
+
+// `args` can be null to describe a raw generic type.
+// For non-generic types it will be zero-length list.
+// TODO: raw types not yet implemented for KE2
 context(KaSession)
 private fun KotlinUsesExtractor.getClassLabel(
     c: KaClassSymbol,
-    argsIncludingOuterClasses: List<Nothing>?
+    argsIncludingOuterClasses: List<List<KaTypeProjection>>?
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/Class.kt:194]: I think it would be worth including a comment with a small example explaining what this list-of-lists means.

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeArgument.kt [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+                    ),
+                    /* OLD: KE1
+                    Unit,
+                    */
+                    wildcardShortName
+                )
+            }
+        }
+        else -> {
+            logger.error("Unexpected type argument.")
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/TypeArgument.kt:65]: Can we include the `.javaClass` in the error please?

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+context(KaSession)
+fun KotlinFileExtractor.extractTypeParameter(
+    tp: KaTypeParameterSymbol,
+    apparentIndex: Int,
+    /* OLD: KE1
+    javaTypeParameter: JavaTypeParameter? */
+): Label<out DbTypevariable>? {
+    with("type parameter", tp) {
+        val parentId = getTypeParameterParentLabel(tp) ?: return null
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt:29]: Should we log an error in the `null` case?

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+import org.jetbrains.kotlin.analysis.api.types.KaTypeParameterType
+import org.jetbrains.kotlin.types.Variance
+
+context(KaSession)
+fun KotlinUsesExtractor.getTypeParameterLabel(param: KaTypeParameterSymbol): String {
+    // Use this instead of `useDeclarationParent` so we can use useFunction with noReplace = true,
+    // ensuring that e.g. a method-scoped type variable declared on kotlin.String.transform<R> gets
+    // a different name to the corresponding java.lang.String.transform<R>, even though
+    // useFunction will usually replace references to one function with the other.
+    val parentLabel = getTypeParameterParentLabel(param)
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt:17]: I think this wants to fail, or possibly return some other best-attempt string, if `parentLabel` is null

File: java/kotlin-extractor2/src/main/kotlin/entities/Types.kt [hunk 15]
```diff
@@ -1,31 +1,30 @@
...
 ): TypeResults {
     // TODO: this cast is unsafe; .symbol is actually a KaClassLikeSymbol
-    val classId = addClassLabel(c.symbol as KaClassSymbol, listOf<Nothing>())
-    val javaResult = TypeResult(classId /* , TODO, TODO */)
+    val javaResult = addClassLabel(c.symbol as KaClassSymbol, c.qualifiers.map { it.typeArguments })
     val kotlinTypeId =
-        tw.getLabelFor<DbKt_class_type>("@\"kt_class;{$classId}\"") {
-            tw.writeKt_class_types(it, classId)
+        tw.getLabelFor<DbKt_class_type>("@\"kt_class;{${javaResult.id}}\"") {
+            tw.writeKt_class_types(it, javaResult.id)
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/Types.kt:19]: I don't think this refactoring is going to do the right thing in the future, e.g. for `Int`, where the Java ID might be for `int` but the Kotlin ID should be for `Int`

File: java/kotlin-extractor2/src/main/kotlin/entities/Types.kt [hunk 16]
```diff
@@ -97,24 +86,24 @@ fun KotlinUsesExtractor.useType(t: KaType?, context: TypeContext = TypeContext.O
                                 val kotlinBaseTypeId = kotlinResultBase.id
                                 val kotlinAliasTypeId =
                                     tw.getLabelFor<DbKt_type_alias>("@\"kt_type_alias;{$classId};{$kotlinBaseTypeId}\"") {
-                                        tw.writeKt_type_aliases(it, classId, kotlinBaseTypeId)
+                                        tw.writeKt_type_aliases(it, classId.id, kotlinBaseTypeId)
```
Comment (igfoo) [inline java/kotlin-extractor2/src/main/kotlin/entities/Types.kt:89]: `classId` should probably be renamed `classResult` or similar.

File: java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt (hunk 10 above)
Comment (smowton) [reply java/kotlin-extractor2/src/main/kotlin/KotlinExtractor.kt:169]: You might want to include the context at https://github.com/github/codeql/pull/18215#issuecomment-2520120125 identifying thread-pool cleanup as the cause

File: java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt (hunk 13 above)
Comment (smowton) [reply java/kotlin-extractor2/src/main/kotlin/entities/TypeParameter.kt:29]: The callee already logs, but I've added error-cascading per this and the comment below.

File: java/kotlin-extractor2/src/main/kotlin/entities/Types.kt (hunk 15 above)
Comment (smowton) [reply java/kotlin-extractor2/src/main/kotlin/entities/Types.kt:19]: Added a TODO -- I wasn't intending to address this here, only to keep the existing behaviour.

**Comment by smowton** [comment]
@igfoo comments addressed

**Review by igfoo — APPROVED** [review]
